infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -608,23 +606,25 @@ private boolean clusterRoleChanged(TopologyDescriptor discoveredTopology) {
      */
     // TODO V2: This method can be renamed to something more generic.  It is the default handling for any topology
     //  change where the role has not changed.  This method assumes that only Sink clusters can change.  Once more
-    //  topologies are supported, Source clusters can also be added/removed.
+    //  topologies are supported, Source clusters can also be added/removed.:::Coming in connectionModel PR
     private void onSinkClusterAddRemove(TopologyDescriptor newTopology) {
         log.debug("Sink Cluster has been added or removed");
 
+        topologyDescriptor = newTopology;
+
+        // refresh the session so new sessions are added/removed.
+        sessionManager.refresh(newTopology);
+
         // Only process new sinks if the local cluster role is SOURCE
-        if (topologyDescriptor.getLocalClusterDescriptor().getRole() == ClusterRole.SOURCE) {
-            sessionManager.refresh(newTopology);
+        if (isSource()) {
             sessionManager.startReplication();
         } else {
```

Comment:
I must have missed this from session manager PR's review, but shouldn't startReplication on leader node only? 

I think for `refresh(topology)`should also be on leader node, otherwise all the node's session manager will try to update metadata tables for the same thing? I think the only thing available for all the nodes is `LogReplicationServer` for querying replication status?

Reply:
Yes, startReplication() should be from the leader node only.  Right now, we do not check this when invoked from onSinkClusterAddRemove().  

The SessionManager maintains session info in-memory and updates the metadata table.  I see the point of updating the tables only from the leader but in-memory view should be with all nodes of the cluster IMO.  Otherwise we should create the SessionManager only after receiving the leadership info, which is not how it is right now.

Reply:
I had seen this while adding the refresh but looks like I missed addressing it.

I'll add the leadership check, and if its not a lot of change, I'll try to have the sessionManager.refresh only for leader. Other wise I'll add a TODO for now and address it in the connectionModel PR, the reason being that all this is going to move around when a cluster can be both source/sink. 

Also, we currently have a silent issue where the metadata gets overwritten by every node in the cluster. Leader refreshing sessions would resolve that issue.